### PR TITLE
Lay ground work for new types of secure transports

### LINKED ISSUE #210: Add `secio`
The repo currently lacks a `secio` implementation; we should add one :)

### PR DESCRIPTION
**Pull request by ralexstokes**
Ground work for adding in `secio` transport.

Next PR on top of this one will add `secio` implementation.

### CODE REVIEW

File: libp2p/security/security_multistream.py [hunk 1]
```diff
@@ -24,64 +21,48 @@
...
-    transports: Dict[TProtocol, "ISecureTransport"]
-    multiselect: "Multiselect"
-    multiselect_client: "MultiselectClient"
+    transports: Dict[TProtocol, ISecureTransport]
+    multiselect: Multiselect
+    multiselect_client: MultiselectClient
 
-    def __init__(self) -> None:
-        # Map protocol to secure transport
+    def __init__(self, secure_transports_by_protocol) -> None:
```
Comment (pipermerriam) [inline libp2p/security/security_multistream.py]: missing type hints?  Any chance we can get things typed and then enforced via CI?

File: tests/utils.py [hunk 2]
```diff
@@ -52,3 +54,8 @@
     # Associate the peer with local ip address (see default parameters of Libp2p())
     node_a.get_peerstore().add_addrs(node_b.get_id(), node_b.get_addrs(), 10)
     return node_a, node_b
+
+
+def generate_new_private_key():
+    new_key = RSA.generate(2048, e=65537)
+    return new_key
```
Comment (pipermerriam) [inline tests/utils.py]: looks identical to the `libp2p.generate_new_private_key` function.  Any reason it's duplicated?

File: libp2p/security/security_multistream.py (hunk 1 above)
Comment (ralexstokes) [reply libp2p/security/security_multistream.py]: yep, there is progress on restoring use of `mypy` in this repo -- i think it has already been completed so I'll have to fix this before we can merge this in.

File: tests/utils.py (hunk 2 above)
Comment (ralexstokes) [reply tests/utils.py]: it is :)

"a little copy is better than a little dependency", esp when the code is short and you are moving fast ;)

i'll make a note to deprecate this (we need a better way to handle key gen anyway)

File: libp2p/__init__.py [hunk 3]
```diff
@@ -93,8 +98,14 @@ def initialize_default_swarm(
     # TODO TransportUpgrader is not doing anything really
     # TODO parse muxer and sec to pass into TransportUpgrader
     muxer = muxer_opt or ["mplex/6.7.0"]
-    sec = sec_opt or {TProtocol("insecure/1.0.0"): InsecureTransport("insecure")}
-    upgrader = TransportUpgrader(sec, muxer)
+    private_key_bytes = private_key.export_key("DER")
```
Comment (mhchia) [inline libp2p/__init__.py]: Just a note: I learned we lack `KeyBook` implementation, which is useful when we need to store peers' public keys in the future. Ref: https://github.com/libp2p/go-libp2p-peerstore/blob/master/pstoremem/keybook.go

File: tests/security/test_security_multistream.py [hunk 4]
```diff
@@ -1,14 +1,13 @@
...
 from libp2p import new_node
 from libp2p.peer.peerinfo import info_from_p2p_addr
 from libp2p.protocol_muxer.multiselect_client import MultiselectClientError
-from libp2p.security.insecure_security import InsecureTransport
-from libp2p.security.simple_security import SimpleSecurityTransport
-from tests.utils import cleanup, connect
+from libp2p.security.insecure.transport import InsecureSession, InsecureTransport
+from libp2p.security.simple.transport import SimpleSecurityTransport
+import multiaddr
+import pytest
```
Comment (mhchia) [inline tests/security/test_security_multistream.py]: Is this sorting(`libp2p` above `multiaddr` and `pytest`) desired by us? If not, we probably need to adjust the configuration of isort a little bit.

**Comment by ralexstokes** [comment]
just a note, this was closed after a rename operation i tried locally... will make new PR

File: tests/security/test_security_multistream.py (hunk 4 above)
Comment (ralexstokes) [reply tests/security/test_security_multistream.py]: yeah i noticed this -- not sure what is going on here

Comment (ralexstokes) [reply tests/security/test_security_multistream.py]: this seems to have magically fixed itself after i rebased master in #250.

File: libp2p/__init__.py (hunk 3 above)
Comment (ralexstokes) [reply libp2p/__init__.py]: yeah let's make an issue

File: libp2p/security/security_multistream.py (hunk 1 above)
Comment (ralexstokes) [reply libp2p/security/security_multistream.py]: this has been addressed in #250.

File: tests/utils.py (hunk 2 above)
Comment (ralexstokes) [reply tests/utils.py]: also fixed in other work, reflected in #250.

**Pull request closed without merging**